nomos-mix/core/src/persistent_transmission.rs

```diff
@@ -26,6 +30,86 @@ impl Default for PersistentTransmissionSettings {
     }
 }
 
+pub struct PersistentTransmissionStream<S>
+where
+    S: Stream,
+{
+    interval: Interval,
+    coin: Coin<ChaCha12Rng>,
+    queue: VecDeque<S::Item>,
+    stream: S,
+}
+
+impl<S> PersistentTransmissionStream<S>
+where
+    S: Stream,
+{
+    pub fn new(
+        settings: PersistentTransmissionSettings,
+        stream: S,
+    ) -> PersistentTransmissionStream<S> {
+        let interval = time::interval(Duration::from_secs_f64(
+            1.0 / settings.max_emission_frequency,
+        ));
+        let coin = Coin::<_>::new(
+            ChaCha12Rng::from_entropy(),
+            settings.drop_message_probability,
+        )
+        .unwrap();
+        Self {
+            interval,
+            coin,
+            queue: Default::default(),
+            stream,
+        }
+    }
+}
+
+impl<S> Stream for PersistentTransmissionStream<S>
+where
+    S: Stream<Item = Vec<u8>> + Unpin,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let Self {
+            ref mut interval,
+            ref mut queue,
+            ref mut stream,
+            ref mut coin,
+            ..
+        } = self.get_mut();
+        if pin!(interval).poll_tick(cx).is_ready() {
+            match queue.pop_front() {
+                item @ Some(_) => return Poll::Ready(item),
+                None => {
+                    if coin.flip() {
+                        return Poll::Ready(Some(DROP_MESSAGE.to_vec()));
+                    }
+                }
+            };
+        };
+        if let Poll::Ready(Some(item)) = pin!(stream).poll_next(cx) {
+            queue.push_back(item);
+        }
+        Poll::Pending
+    }
+}
+
+pub trait PersistentTransmissionExt: Stream {
+    fn persistent_transmission(
+        self,
+        settings: PersistentTransmissionSettings,
+    ) -> PersistentTransmissionStream<Self>
+    where
+        Self: Sized + Unpin,
+    {
+        PersistentTransmissionStream::new(settings, self)
+    }
+}
+
+impl<S> PersistentTransmissionExt for S where S: Stream {}
```

Comment:
Oh, I haven't known about this skill, which implements `PersistentTransmissionExt` for any type that implements `Stream`.
If this is a common practice, I think I can adopt this for the `TemporalProcessor` that implements `Stream`, to have the common format.